### PR TITLE
Fix undefined parameter

### PR DESCRIPTION
**Pull request by czpmango**
## What type of PR is this?
- [x] bug
- [ ] feature
- [ ] enhancement

#### Issue(s) number: 
https://github.com/vesoft-inc/nebula-ent/issues/2485

## Checklist:
Tests:
- [ ] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Performance test
- [ ] N/A

Affects:
- [x] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [x] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

### CODE REVIEW

File: src/graph/util/ExpressionUtils.cpp [hunk 1]
```diff
@@ -248,6 +250,43 @@ Expression *ExpressionUtils::rewriteParameter(const Expression *expr, QueryConte
...
+           !ve->isInner();
+  };
+
+  if (finder(expr)) {
+    return {static_cast<const VariableExpression *>(expr)->var()};
+  }
+  FindVisitor visitor(finder);
+  const_cast<Expression *>(expr)->accept(&visitor);
+  auto varExprs = visitor.results();
+  std::vector<std::string> vars;
```
Comment (Aiee) [inline src/graph/util/ExpressionUtils.cpp:283]: allocate the size in the initialization

File: src/graph/util/ExpressionUtils.cpp [hunk 2]
```diff
@@ -4,6 +4,8 @@
 
 #include "graph/util/ExpressionUtils.h"
 
+#include <vector>
```
Comment (Aiee) [inline src/graph/util/ExpressionUtils.cpp:7]: Remove

File: src/graph/validator/MatchValidator.cpp [hunk 3]
```diff
@@ -334,6 +334,10 @@ Status MatchValidator::buildEdgeInfo(const MatchPath *path,
 // Rewrite expression to fit semantic, check type and check used aliases.
 Status MatchValidator::validateFilter(const Expression *filter,
                                       WhereClauseContext &whereClauseCtx) {
+  auto vars = graph::ExpressionUtils::ExtractInnerVars(filter, qctx_);
+  if (!vars.empty()) {
+    return Status::SemanticError("Undefined parameter: " + vars[0]);
```
Comment (Aiee) [inline src/graph/validator/MatchValidator.cpp:339]: Why print only the first parameter?

**Review by jievince — DISMISSED** [review]

**Review by yixinglu — DISMISSED** [review]
LGTM

File: src/graph/util/ExpressionUtils.cpp (hunk 1 above)
Comment (czpmango) [reply src/graph/util/ExpressionUtils.cpp:283]: Not necessary.

File: src/graph/util/ExpressionUtils.cpp (hunk 2 above)
Comment (czpmango) [reply src/graph/util/ExpressionUtils.cpp:7]: ACK.

**Review by jievince — APPROVED** [review]

File: src/graph/validator/MatchValidator.cpp (hunk 3 above)
Comment (Aiee) [reply src/graph/validator/MatchValidator.cpp:339]: If there are multiple undefined parameters then the user needs to edit the query multiple times to find all parameters. Why is printing them all at once not user-friendly?

Comment (czpmango) [reply src/graph/validator/MatchValidator.cpp:339]: ACK.

**Review by jievince — APPROVED** [review]

File: src/graph/validator/LookupValidator.cpp [hunk 4]
```diff
@@ -192,6 +192,21 @@ Status LookupValidator::validateWhere() {
...
+        undefinedParams.emplace_back(var);
+      }
+    }
+    if (!undefinedParams.empty()) {
+      return Status::SemanticError(
+          "Undefined parameters: " +
+          std::accumulate(++undefinedParams.begin(),
+                          undefinedParams.end(),
+                          *undefinedParams.begin(),
+                          [](auto& lhs, auto& rhs) { return lhs + ", " + rhs; }));
```
[the remark this answers is not in the document]
Comment (czpmango) [reply src/graph/validator/LookupValidator.cpp:208]: `std::accumulate` is also okey, folly lib sounds a little heavy.

**Pull request merged**